pkg/controllers/provisioningv2/harvestercleanup/controller.go

```diff
@@ -0,0 +1,165 @@
+	if obj == nil {
+		return obj, nil
+	}
+
+	objMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// Exit immediately if the object is not being deleted.
+	if objMeta.GetDeletionTimestamp().IsZero() {
+		return obj, nil
+	}
+
+	objData, err := data.Convert(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	key := objMeta.GetNamespace() + "/" + objMeta.GetName()
+
+	capiCluster, err := capr.GetCAPIClusterFromLabel(obj, h.capiClusters)
+	if apierrors.IsNotFound(err) {
+		logrus.Warnf("[%s] %s: there is no cluster (apiversion=cluster.x-k8s.io). Continue ...", handlerName, key)
+		return obj, nil
+	}
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting cluster (apiversion=cluster.x-k8s.io): %v", handlerName, key, err)
+		return obj, err
+	}
+
+	// Skip if the CAPI cluster (cluster.x-k8s.io) is not being deleted.
+	if capiCluster.DeletionTimestamp.IsZero() {
+		return obj, nil
+	}
+
+	cluster, err := h.clusterCache.Get(objMeta.GetNamespace(), capiCluster.Name)
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting cluster (apiversion=provisioning.cattle.io): %v", handlerName, key, err)
+		return obj, err
+	}
+
+	secret, err := machineprovision.GetCloudCredentialSecret(h.secretCache, "", cluster.Spec.CloudCredentialSecretName)
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting cloud credential secret: %v", handlerName, key, err)
+		return obj, err
+	}
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(secret.Data["harvestercredentialConfig-kubeconfigContent"])
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting REST config from cloud credential secret: %v", handlerName, key, err)
+		return obj, err
+	}
+
+	// We need to use the dynamic client here because we do not want to import
+	// the kubevirt API into Rancher.
+	dynamicClient, err := k8dynamic.NewForConfig(restConfig)
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting dynamic client: %v", handlerName, key, err)
+		return obj, err
+	}
+
+	resourceGVR := schema.GroupVersionResource{
+		Group: "kubevirt.io", Version: "v1", Resource: "virtualmachines",
+	}
+	resourceName := objMeta.GetName()
+	resourceNamespace := objData.String("spec", "vmNamespace")
+
+	resource, err := dynamicClient.Resource(resourceGVR).Namespace(resourceNamespace).Get(context.TODO(), resourceName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		logrus.Warnf("[%s] %s: there is no VM (gvr=%s, namespace=%s, name=%s). Continue ...",
+			handlerName, key, resourceGVR.String(), resourceNamespace, resourceName)
+		return obj, nil
+	}
+	if err != nil {
+		logrus.Errorf("[%s] %s: error getting VM (gvr=%s, namespace=%s, name=%s): %v",
+			handlerName, key, resourceGVR.String(), resourceNamespace, resourceName, err)
+		return obj, err
+	}
+
+	annotations := resource.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	if _, ok := annotations[removedAllPVCsAnnotationKey]; !ok {
+		annotations[removedAllPVCsAnnotationKey] = "true"
```

Comment:
@votdev Could you also consider to add some addtional annotations like guest cluster name/namespace/id/project...

With such annotation on VM, besides this PR https://github.com/harvester/docker-machine-driver-harvester/pull/64, 

Harvester can then add onRemove controller upon VM object, and clean 
 additional resources like LBs besides PVC. thanks.

Reply:
There is a already PR in the pipeline that will take care that the cluster name is set on such VMs. See https://github.com/rancher/rancher/pull/48010.

I think this is the better approach than adding such annotations with this PR here because the mentioned annotations have absolutely nothing to do with what this PR is trying to solve.